crates/sui-core/src/overload_monitor.rs

```diff
@@ -645,4 +683,60 @@ mod tests {
         assert!(0.4 < dropped_ratio);
         assert!(dropped_ratio < 0.6);
     }
+
+    // Tests that the ratio of rejected transactions created randomly matches load shedding percentage in
+    // the overload monitor.
+    #[test]
+    fn test_txn_rejection_rate() {
+        for rejection_percentage in 0..=100 {
+            let mut reject_count = 0;
+            for _ in 0..10000 {
+                let digest = TransactionDigest::random();
+                if should_reject_tx(rejection_percentage, digest, 28455473) {
+                    reject_count += 1;
+                }
+            }
+
+            // Give it a 2% fluctuation.
+            assert!(rejection_percentage as f32 / 100.0 - 0.02 < reject_count as f32 / 10000.0);
+            assert!(reject_count as f32 / 10000.0 < rejection_percentage as f32 / 100.0 + 0.02);
+        }
+    }
+
+    // Tests that rejected transaction will have a chance to be accepted in the future.
+    #[test]
+    #[cfg_attr(msim, ignore)]
+    fn test_txn_rejection_over_time() {
+        let start_time = Instant::now();
+        let mut digest = TransactionDigest::random();
+        let mut minutes_since_epoch = 28455473;
+        let load_shedding_percentage = 50;
+
+        // Find a rejected transaction with 50% rejection rate.
+        while !should_reject_tx(load_shedding_percentage, digest, minutes_since_epoch)
+            && start_time.elapsed() < Duration::from_secs(30)
+        {
+            digest = TransactionDigest::random();
+        }
+
+        // It should always be rejected in the current minute.
+        for _ in 0..100 {
+            assert!(should_reject_tx(
+                load_shedding_percentage,
+                digest,
+                minutes_since_epoch
+            ));
+        }
+
+        // It will be accepted in the future.
+        minutes_since_epoch += 1;
+        while should_reject_tx(load_shedding_percentage, digest, minutes_since_epoch)
+            && start_time.elapsed() < Duration::from_secs(30)
+        {
+            minutes_since_epoch += 1;
+        }
+
+        // Make sure that the tests can finish within 30 seconds.
+        assert!(start_time.elapsed() < Duration::from_secs(30));
```

Comment:
if you make this a simtest you can have a deterministic clock. the test will run fast and be repeatable.